### PR TITLE
lexer: Remove `match-at` dependency, use RegExp

### PR DESCRIPTION
**Pull request by ylemkimon**
Closes #1443.

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -4154,9 +4154,9 @@
       "dev": true
     },
     "flow-bin": {
-      "version": "0.74.0",
-      "resolved": "https://registry.npmjs.org/flow-bin/-/flow-bin-0.74.0.tgz",
-      "integrity": "sha512-tIN9J5qg71S4UbofCu80tve8a+p7Hj7ytwUtu79cLg9KJVVTNnVVJXKgCghVzaZT1Rvl9SMHVPlDs9uYhPHEGQ==",
+      "version": "0.75.0",
+      "resolved": "https://registry.npmjs.org/flow-bin/-/flow-bin-0.75.0.tgz",
+      "integrity": "sha1-uW0e6Z07RGoyJr5mtAEyJM6d8mA=",
```
Comment (edemaine) [inline package-lock.json]: I believe the latest NPM uses `sha512` hashes exclusively. Or does it normally use a mix? I've found I had to delete `node_modules` and clean the NPM cache, then `npm install`, to fix this.

File: src/Lexer.js [hunk 2]
```diff
@@ -67,33 +62,30 @@ const commentRegex = new RegExp(`^${commentRegexString}`);
...
      */
     lex(): Token {
         const input = this.input;
-        const pos = this.pos;
+        const pos = this.tokenRegex.lastIndex;
         if (pos === input.length) {
             return new Token("EOF", new SourceLocation(this, pos, pos));
         }
-        const match = matchAt(tokenRegex, input, pos);
+        const match = this.tokenRegex.exec(input);
```
Comment (edemaine) [inline src/Lexer.js]: I can think of one important difference here: if there's something that doesn't match the regex, `exec` will (I think) skip over it, while `matchAt` will fail.  Thoughts on how to fix this?

File: src/Lexer.js [hunk 3]
```diff
@@ -67,33 +62,30 @@ const commentRegex = new RegExp(`^${commentRegexString}`);
...
 export default class Lexer implements LexerInterface {
     input: string;
-    pos: number;
+    tokenRegex: RegExp;
 
     constructor(input: string) {
         // Separate accents from characters
         this.input = input;
-        this.pos = 0;
+        this.tokenRegex = new RegExp(tokenRegexString, 'g');
```
Comment (edemaine) [inline src/Lexer.js]: It's cool that this works even when `feed`ing another string, thanks to the semantics of `exec`.

File: package-lock.json (hunk 1 above)
Comment (ylemkimon) [reply package-lock.json]: @edemaine There are issues of `npm` mixing integrity hashes, https://github.com/npm/npm/issues/16938 and https://github.com/npm/npm/issues/17749, closed but people are still reporting problems.

File: src/Lexer.js (hunk 2 above)
Comment (ylemkimon) [reply src/Lexer.js]: @edemaine `matchAt` and `exec` both return `null` when match fails, and a `ParseError` will be thrown below.

Comment (kevinbarabash) [reply src/Lexer.js]: If there's a match that occurs after part of the string which isn't matched by the regex, `exec` will happily jump to the next match and return that.  We could keep track of the current position and then verify that match that `exec` occurs at that position.

Comment (ylemkimon) [reply src/Lexer.js]: @kevinbarabash Ah, I didn't know, hence the name match _At_. Added an assertion to check match index matches the last index. Also updated a test case to test regression.

File: package-lock.json (hunk 1 above)
Comment (edemaine) [reply package-lock.json]: Shouldn't we try to get sha512 hashes in there?  Again, I've found resetting NPM as listed above (thanks to these various issues) has fixed the problem.

File: src/Lexer.js [hunk 4]
```diff
@@ -67,33 +62,30 @@ const commentRegex = new RegExp(`^${commentRegexString}`);
...
         const input = this.input;
-        const pos = this.pos;
+        const pos = this.tokenRegex.lastIndex;
         if (pos === input.length) {
             return new Token("EOF", new SourceLocation(this, pos, pos));
         }
-        const match = matchAt(tokenRegex, input, pos);
-        if (match === null) {
+        const match = this.tokenRegex.exec(input);
+        if (match === null || match.index !== pos) {
```
Comment (edemaine) [inline src/Lexer.js]: Nice change.

File: src/Lexer.js (hunk 3 above)
Comment (edemaine) [reply src/Lexer.js]: Oops, I was wrong about this.  I just tried the following in Chrome:

```js
s='aaaaaaa'
re.exec(s)  // ["a", index: 0, input: "aaaaaaa", groups: undefined]
re.exec(s)  // ["a", index: 1, input: "aaaaaaa", groups: undefined]
re.exec(s)  // ["a", index: 2, input: "aaaaaaa", groups: undefined]
t='aaaa'
re.exec(t)  // ["a", index: 3, input: "aaaa", groups: undefined]
```

It seems we should set `this.tokenRegex.lastIndex = 0` in `feed`.

Comment (ylemkimon) [reply src/Lexer.js]: @edemaine Is it possible to feed new string to the lexer? Currently, new `RegExp` instance is created when instantiating lexer.

Comment (kevinbarabash) [reply src/Lexer.js]: I don't believe so.

File: test/errors-spec.js [hunk 5]
```diff
@@ -285,9 +285,9 @@ describe("Lexer:", function() {
 
     describe("#_innerLex", function() {
         it("rejects lone surrogate char", function() {
-            expect("\udcba").toFailWithParseError(
+            expect("\udcba ").toFailWithParseError(
                    "Unexpected character: '\udcba' at position 1:" +
-                    " \udcba\u0332");
+                    " \udcba\u0332 ");
```
Comment (kevinbarabash) [inline test/errors-spec.js]: Is the addition o the space necessary in this test?

File: src/Lexer.js (hunk 3 above)
Comment (edemaine) [reply src/Lexer.js]: Oops, indeed, all is well. I thought `Lexer` had the `feed` method but it's `MacroExpander`, which creates a new `Lexer` each time.

**Comment by edemaine** [comment]
I'm curious whether this speeds up KaTeX at all. Probably too marginal to tell.

**Pull request merged**